File: pydss/conversion/PCElementImpl.py
from dss.conversion.PCElement import (PCElement,)
from dss.common.impl.DSSGlobals import (DSSGlobals,)
from dss.common.impl.DSSClassDefs import (DSSClassDefs,)
from dss.common.impl.DSSCktElement import (DSSCktElement,)
# from org.apache.commons.math.complex.Complex import (Complex,)


class PCElementImpl(DSSCktElement, PCElement):
    ITerminalUpdated = None
    spectrum = None
    spectrumObj = None
    meterObj = None
    sensorObj = None
    injCurrent = None

    def __init__(self, ParClass):
        super(PCElementImpl, self)(ParClass)
        self.spectrum = 'default'
        self.spectrumObj = None
        # have to allocate later because not guaranteed there will be one now
        self.sensorObj = None
        self.meterObj = None
        self.injCurrent = None
        self.ITerminalUpdated = False
        self.objType = DSSClassDefs.PC_ELEMENT

    def injCurrents(self):
        """Add injection currents into system currents array."""
        sol = DSSGlobals.activeCircuit.getSolution()
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.YOrder):
                break
            sol.getCurrents()[self.nodeRef[i]] = sol.getCurrents()[self.nodeRef[i]].add(self.injCurrent[i])
        return 0

    def getInjCurrents(self, curr):
        """Get present values of terminal."""
        DSSGlobals.doErrorMsg('PCElement.InjCurrents', 'Improper call to getInjCurrents for element: ' + self.getName() + '.', 'Called PCElement class virtual function instead of actual.', 640)

    def getTerminalCurrents(self, curr):
        """This is called only if we need to compute the terminal currents from the inj currents.

        Such as for harmonic model.
        """
        if self.getITerminalUpdated():
            # just copy ITerminal unless ITerminal=curr
            if curr != self.getITerminal():
                _0 = True
                i = 0
                while True:
                    if _0 is True:
                        _0 = False
                    else:
                        i += 1
                    if not (i < self.YOrder):
                        break
                    curr[i] = self.getITerminal()[i]
        else:
            self.YPrim.vMult(curr, self.getVTerminal())
            _1 = True
            i = 0
            while True:
                if _1 is True:
                    _1 = False
                else:
                    i += 1
                if not (i < self.YOrder):
                    break
                curr[i] = curr[i].add(self.getInjCurrent()[i].negate())
            self.setITerminalUpdated(True)
        self.ITerminalSolutionCount = DSSGlobals.activeCircuit.getSolution().getSolutionCount()

    def getCurrents(self, curr):
        """Get present values of terminal.

        Gets total currents going into a devices terminals.
        """
        try:
            sol = DSSGlobals.activeCircuit.getSolution()
            if self.isEnabled():
                if (
                    sol.lastSolutionWasDirect() and not (sol.isDynamicModel() or sol.isHarmonicModel())
                ):
                    # take a short cut and get currents from YPrim only
                    # for case where model is entirely in Y matrix
                    self.calcYPrimContribution(curr)
                else:
                    self.getTerminalCurrents(curr)
            else:
                # not enabled
                _0 = True
                i = 0
                while True:
                    if _0 is True:
                        _0 = False
                    else:
                        i += 1
                    if not (i < self.YOrder):
                        break
                    curr[i] = Complex.ZERO
        except Exception, e:
            DSSGlobals.doErrorMsg('getCurrents for element: ' + self.getName() + '.', e.getMessage(), 'Inadequate storage allotted for circuit element.', 641)

    def calcYPrimContribution(self, curr):
        self.computeVTerminal()
        # apply these voltages to Yprim
        self.YPrim.vMult(curr, self.VTerminal)

    def initHarmonics(self):
        """For harmonics mode"""
        # by default do nothing in the base class
        pass

    def initPropertyValues(self, arrayOffset):
        self.propertyValue[arrayOffset + 1] = self.spectrum
        super(PCElementImpl, self).initPropertyValues(arrayOffset + 1)

    def initStateVars(self):
        """For dynamics mode and control devices."""
        # by default do nothing
        pass

    def integrateStates(self):
        # by default do nothing
        pass

    def getAllVariables(self, states):
        # Do nothing
        pass

    def numVariables(self):
        return 0

    def variableName(self, i):
        # Do nothing
        return ''

    def lookupVariable(self, s):
        """Search through variable name list and return index if found.
        Compare up to length of S.
        """
        result = -1
        # returns -1 for error not found
        testLength = len(s)
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.numVariables()):
                break
            if self.variableName(i)[:testLength].equalsIgnoreCase(s):
                result = i
                break
        return result

    def dumpProperties(self, f, complete):
        super(PCElementImpl, self).dumpProperties(f, complete)
        if complete:
            print '! VARIABLES'
            _0 = True
            i = 0
            while True:
                if _0 is True:
                    _0 = False
                else:
                    i += 1
                if not (i < self.numVariables()):
                    break
                print '! ' + i + ': ' + self.variableName(i) + ' = ' + String.format.format('%-.5g', self.getVariable(i))

    def getVariable(self, i):
        # Do nothing here -- up to override function
        return -9999.99

    def setVariable(self, i, value):
        # Do nothing
        pass

    def computeITerminal(self):
        ckt = DSSGlobals.activeCircuit
        if self.ITerminalSolutionCount != ckt.getSolution().getSolutionCount():
            self.getCurrents(self.ITerminal)
            self.ITerminalSolutionCount = ckt.getSolution().getSolutionCount()

    def zeroInjCurrent(self):
        _0 = True
        i = 0
        while True:
            if _0 is True:
                _0 = False
            else:
                i += 1
            if not (i < self.YOrder):
                break
            self.injCurrent[i] = Complex.ZERO

    def setITerminalUpdated(self, value):
        self.ITerminalUpdated = value
        if value:
            self.ITerminalSolutionCount = DSSGlobals.activeCircuit.getSolution().getSolutionCount()

    def getITerminalUpdated(self):
        return self.ITerminalUpdated

    def getInjCurrent(self):
        return self.injCurrent

    def getSpectrum(self):
        return self.spectrum

    def setSpectrum(self, value):
        self.spectrum = value

    def getSpectrumObj(self):
        """Upline sensor for this element"""
        return self.spectrumObj

    def setSpectrumObj(self, spectrum):
        self.spectrumObj = spectrum

    def getMeterObj(self):
        """Upline energy meter"""
        return self.meterObj

    def setMeterObj(self, meter):
        self.meterObj = meter

    def getSensorObj(self):
        return self.sensorObj

    def setSensorObj(self, sensor):
        self.sensorObj = sensor

    def setInjCurrent(self, current):
        self.injCurrent = current

File: core/dss-core/src/main/java/com/ncond/dss/conversion/impl/PCElementImpl.java
package com.ncond.dss.conversion.impl;

import java.io.PrintStream;

import org.apache.commons.math.complex.Complex;

import com.ncond.dss.common.Circuit;
import com.ncond.dss.common.DSSClass;
import com.ncond.dss.common.SolutionObj;
import com.ncond.dss.common.impl.CktElementImpl;
import com.ncond.dss.common.impl.DSSClassDefs;
import com.ncond.dss.common.impl.DSS;
import com.ncond.dss.conversion.PCElement;
import com.ncond.dss.general.SpectrumObj;
import com.ncond.dss.meter.MeterElement;

public abstract class PCElementImpl extends CktElementImpl implements PCElement {

	private boolean ITerminalUpdated;

	protected String spectrum;
	protected SpectrumObj spectrumObj;
	/** Upline energy meter */
	private MeterElement meterObj;
	/** Upline sensor for this element */
	private MeterElement sensorObj;

	private Complex[] injCurrent;

	public PCElementImpl(DSSClass ParClass) {
		super(ParClass);
		spectrum = "default";
		spectrumObj = null;  // have to allocate later because not guaranteed there will be one now
		sensorObj   = null;
		meterObj    = null;
		injCurrent  = null;
		ITerminalUpdated = false;

		objType = DSSClassDefs.PC_ELEMENT;
	}

	/**
	 * Add injection currents into system currents array.
	 */
	@Override
	public int injCurrents() {
		SolutionObj sol = DSS.activeCircuit.getSolution();

		for (int i = 0; i < YOrder; i++)
			sol.setCurrent(nodeRef[i], sol.getCurrent(nodeRef[i]).add( injCurrent[i] ));

		return 0;
	}

	/**
	 * Get present values of terminal.
	 */
	@Override
	public void getInjCurrents(Complex[] curr) {
		DSS.doErrorMsg("PCElement.InjCurrents", ("Improper call to getInjCurrents for element: " + getName() + "."),
			"Called PCElement class virtual function instead of actual.", 640);
	}

	/**
	 * This is called only if we need to compute the terminal currents from the inj currents.
	 *
	 * Such as for harmonic model.
	 */
	protected void getTerminalCurrents(Complex[] curr) {
		if (getITerminalUpdated()) {  // just copy ITerminal unless ITerminal=curr
			if (curr != getITerminal())
				for (int i = 0; i < YOrder; i++)
					curr[i] = getITerminal()[i];
		} else {
			YPrim.vMult(curr, getVTerminal());
			for (int i = 0; i < YOrder; i++)
				curr[i] = curr[i].add( getInjCurrent()[i].negate() );
			setITerminalUpdated(true);
		}
		ITerminalSolutionCount = DSS.activeCircuit.getSolution().getSolutionCount();
	}

	/**
	 * Get present values of terminal.
	 *
	 * Gets total currents going into a devices terminals.
	 */
	@Override
	public void getCurrents(Complex[] curr) {
		try {
			SolutionObj sol = DSS.activeCircuit.getSolution();

			if (isEnabled()) {

				if ( sol.lastSolutionWasDirect() && (! (sol.isDynamicModel() || sol.isHarmonicModel()) ) ) {

					// take a short cut and get currents from YPrim only
					// for case where model is entirely in Y matrix

					calcYPrimContribution(curr);

				} else {

					getTerminalCurrents(curr);
				}

			} else {  // not enabled
				for (int i = 0; i < YOrder; i++)
					curr[i] = Complex.ZERO;
			}
		} catch (Exception e) {
			DSS.doErrorMsg(("getCurrents for element: " + getName() + "."), e.getMessage(),
					"Inadequate storage allotted for circuit element.", 641);
		}
	}

	@Override
	public void calcYPrimContribution(Complex[] curr) {
		computeVTerminal();
		// apply these voltages to Yprim
		YPrim.vMult(curr, VTerminal);
	}

	/**
	 * For harmonics mode
	 */
	@Override
	public void initHarmonics() {
		// by default do nothing in the base class
	}

	@Override
	public void initPropertyValues(int arrayOffset) {
		propertyValue[arrayOffset + 1] = spectrum;

		super.initPropertyValues(arrayOffset + 1);
	}

	/**
	 * For dynamics mode and control devices.
	 */
	@Override
	public void initStateVars() {
		// by default do nothing
	}

	@Override
	public void integrateStates() {
		// by default do nothing
	}

	@Override
	public void getAllVariables(double[] states) {
		/* Do nothing */
	}

	@Override
	public int numVariables() {
		return 0;
	}

	@Override
	public String variableName(int i) {
		/* Do nothing */
		return "";
	}

	/**
	 * Search through variable name list and return index if found.
	 * Compare up to length of S.
	 */
	@Override
	public int lookupVariable(String s) {
		int result = -1;   // returns -1 for error not found
		int testLength = s.length();
		for (int i = 0; i < numVariables(); i++) {
			if (variableName(i).substring(0, testLength).equalsIgnoreCase(s)) {
				result = i;
				break;
			}
		}
		return result;
	}

	@Override
	public void dumpProperties(PrintStream f, boolean complete) {
		super.dumpProperties(f, complete);

		if (complete) {
			f.println("! VARIABLES");
			for (int i = 0; i < numVariables(); i++)
				f.println("! " + i + ": " + variableName(i) + " = " + String.format("%-.5g", getVariable(i)));
		}
	}

	@Override
	public double getVariable(int i) {
		/* Do nothing here -- up to override function */
		return -9999.99;
	}

	@Override
	public void setVariable(int i, double value) {
		/* Do nothing */
	}

	@Override
	public void computeITerminal() {
		Circuit ckt = DSS.activeCircuit;

		if (ITerminalSolutionCount != ckt.getSolution().getSolutionCount()) {
			getCurrents(ITerminal);
			ITerminalSolutionCount = ckt.getSolution().getSolutionCount();
		}
	}

	@Override
	public void zeroInjCurrent() {
		for (int i = 0; i < YOrder; i++)
			injCurrent[i] = Complex.ZERO;
	}

	@Override
	public void setITerminalUpdated(boolean value) {
		ITerminalUpdated = value;
		if (value)
			ITerminalSolutionCount = DSS.activeCircuit.getSolution().getSolutionCount();
	}

	@Override
	public boolean getITerminalUpdated() {
		return ITerminalUpdated;
	}

	@Override
	public Complex[] getInjCurrent() {
		return injCurrent;
	}

	@Override
	public String getSpectrum() {
		return spectrum;
	}

	@Override
	public void setSpectrum(String value) {
		spectrum = value;
	}

	/** Upline sensor for this element */
	@Override
	public SpectrumObj getSpectrumObj() {
		return spectrumObj;
	}

	@Override
	public void setSpectrumObj(SpectrumObj spectrum) {
		spectrumObj = spectrum;
	}

	/** Upline energy meter */
	@Override
	public MeterElement getMeterObj() {
		return meterObj;
	}

	@Override
	public void setMeterObj(MeterElement meter) {
		meterObj = meter;
	}

	@Override
	public MeterElement getSensorObj() {
		return sensorObj;
	}

	@Override
	public void setSensorObj(MeterElement sensor) {
		sensorObj = sensor;
	}

	@Override
	public void setInjCurrent(Complex[] current) {
		injCurrent = current;
	}

}
